### PR TITLE
[file_selector] Update file_selector_platform_interface by adding the new property 'uniformTypeIdentifiers' to the XTypeGroup.

### PR DESCRIPTION
**Pull request by juandausa**
## Update file_selector_platform_interface by adding the new property 'uniformTypeIdentifiers' to the XTypeGroup.

- Update file_selector_platform_interface by adding the new property 'uniformTypeIdentifiers'.

*List which issues are fixed by this PR. You must list at least one issue.*
- [Rename XTypeGroup's macUTIs #103743](https://github.com/flutter/flutter/issues/103743)

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/plugins/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/main/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/file_selector/file_selector_platform_interface/lib/src/types/x_type_group/x_type_group.dart [hunk 1]
```diff
@@ -26,7 +26,7 @@ class XTypeGroup {
   final List<String>? mimeTypes;
 
   /// The UTIs for this group
-  final List<String>? macUTIs;
+  List<String>? macUTIs;
```
Comment (stuartmorgan) [inline packages/file_selector/file_selector_platform_interface/lib/src/types/x_type_group/x_type_group.dart:29]: This should stay final; there shouldn't be a setter for it. I forgot this was a conceptually immutable class when I wrote that example of forwarding constructs in the issue.

File: packages/file_selector/file_selector_platform_interface/lib/src/types/x_type_group/x_type_group.dart [hunk 2]
```diff
@@ -50,6 +50,14 @@ class XTypeGroup {
         (webWildCards?.isEmpty ?? true);
   }
 
+  /// Returns the list of Uniform Type Identifiers for this group.
+  List<String>? get uniformTypeIdentifiers => macUTIs;
```
Comment (stuartmorgan) [inline packages/file_selector/file_selector_platform_interface/lib/src/types/x_type_group/x_type_group.dart:54]: The field should be renamed, and the aliasing be done in the other direction, because the alias will be deprecated later, not the field.

File: packages/file_selector/file_selector_platform_interface/test/x_type_group_test.dart [hunk 3]
```diff
@@ -7,12 +7,13 @@ import 'package:flutter_test/flutter_test.dart';
...
+    final List<String> mimeTypes = <String>['text/plain'];
+    final List<String> macUTIs = <String>['public.plain-text'];
+    final List<String> webWildCards = <String>['image/*'];
+
     test('toJSON() creates correct map', () {
       const String label = 'test group';
-      final List<String> extensions = <String>['txt', 'jpg'];
-      final List<String> mimeTypes = <String>['text/plain'];
-      final List<String> macUTIs = <String>['public.plain-text'];
-      final List<String> webWildCards = <String>['image/*'];
```
Comment (stuartmorgan) [inline packages/file_selector/file_selector_platform_interface/test/x_type_group_test.dart:15]: Please revert this change, and just inline whatever values you need in the new tests. Tests should generally be as localized as possible, to make them easy to understand and change in isolation.

File: packages/file_selector/file_selector_platform_interface/lib/src/types/x_type_group/x_type_group.dart (hunk 1 above)
Comment (stuartmorgan) [reply packages/file_selector/file_selector_platform_interface/lib/src/types/x_type_group/x_type_group.dart:29]: (Filed https://github.com/flutter/flutter/issues/111906 since this change shouldn't have been possible in the first place.)

File: packages/file_selector/file_selector_platform_interface/lib/src/types/x_type_group/x_type_group.dart (hunk 2 above)
Comment (stuartmorgan) [reply packages/file_selector/file_selector_platform_interface/lib/src/types/x_type_group/x_type_group.dart:54]: Relatedly, the constructor needs to be updated to allow passing `uniformTypeIdentifiers`.

The behavior needs to be that we `assert` that they aren't both non-null, and then the field is set to `uniformTypeIdentifiers ?? macUTIs`.

File: packages/file_selector/file_selector_platform_interface/test/x_type_group_test.dart [hunk 4]
```diff
@@ -71,6 +72,24 @@ void main() {
...
+    });
+
+    test('getUniformTypeIdentifiers returns macUTIs', () {
+      final XTypeGroup group = XTypeGroup(
+        macUTIs: macUTIs,
+      );
+
+      expect(group.uniformTypeIdentifiers, macUTIs);
+    });
+
```
Comment (stuartmorgan) [inline packages/file_selector/file_selector_platform_interface/test/x_type_group_test.dart:92]: The constructor changes requested above will need new tests as well.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
Thanks for the submission!

File: packages/file_selector/file_selector_platform_interface/CHANGELOG.md [hunk 5]
```diff
@@ -1,5 +1,6 @@
-## NEXT
+## 2.1.1
```
Comment (stuartmorgan) [inline packages/file_selector/file_selector_platform_interface/CHANGELOG.md:1]: Oh, and this should actually be 2.2.0 since it's new API.

File: packages/file_selector/file_selector_platform_interface/lib/src/types/x_type_group/x_type_group.dart (hunk 2 above)
Comment (juandausa) [reply packages/file_selector/file_selector_platform_interface/lib/src/types/x_type_group/x_type_group.dart:54]: Hi Stuart, currently the constructor allows null values for the `macUTIs` field. Do you mean that if both `macUTIs` and `uniformTypeIdentifiers` are null, we should just throw an exception? Wouldn't it be a breaking change?

Comment (juandausa) [reply packages/file_selector/file_selector_platform_interface/lib/src/types/x_type_group/x_type_group.dart:54]: I've been thinking about this, and now I understand that you meant if both are not null we should throw an exception, shouldn't we?

Comment (stuartmorgan) [reply packages/file_selector/file_selector_platform_interface/lib/src/types/x_type_group/x_type_group.dart]: Yes, that's correct.

**Comment by adpinola** [comment]
Hi @stuartmorgan we introduced the changes you requested, can you take a look again? Thanks.

File: packages/file_selector/file_selector_platform_interface/lib/src/types/x_type_group/x_type_group.dart [hunk 6]
```diff
@@ -14,18 +14,22 @@ class XTypeGroup {
...
     List<String>? extensions,
     this.mimeTypes,
-    this.macUTIs,
+    List<String>? macUTIs,
+    List<String>? uniformTypeIdentifiers,
     this.webWildCards,
-  }) : _extensions = extensions;
+  })  : _extensions = extensions,
+        assert(uniformTypeIdentifiers == null || macUTIs == null,
+            'It is only allowed to specify either macUTIs or uniformTypeIdentifiers'),
```
Comment (stuartmorgan) [inline packages/file_selector/file_selector_platform_interface/lib/src/types/x_type_group/x_type_group.dart:22]: "Only one of uniformTypeIdentifiers or macUTIs can be non-null"

**Comment by adpinola** [comment]
Feedback applied.
Moving forward to replacing the old `macUTIs`, we will ` @Deprecate` it, and then we should replace it with the new `uniformTypeIdentifiers`. 
Can this be done in a single PR for all platforms, or we should create one to mark the property as deprecated and then replace it in one PR for each platform?

Thanks.

**Review by stuartmorgan — APPROVED** [review]
LGTM, @ditman for second review.

**Comment by stuartmorgan** [comment]
> Moving forward to replacing the old `macUTIs`, we will ` @Deprecate` it, and then we should replace it with the new `uniformTypeIdentifiers`. Can this be done in a single PR for all platforms, or we should create one to mark the property as deprecated and then replace it in one PR for each platform?

The order is:
- publish the new API
- update all of the code in flutter/plugins to use the new API instead of the old one
- deprecate the old API

The last step can't be done before the previous one because it will break our CI analysis.

File: packages/file_selector/file_selector_platform_interface/test/x_type_group_test.dart [hunk 7]
```diff
@@ -71,7 +70,70 @@ void main() {
...
+        'public.plain-images'
+      ];
+      expect(
+          () => XTypeGroup(
+              macUTIs: macUTIs,
+              uniformTypeIdentifiers: uniformTypeIndentifiers),
+          throwsA(predicate((Object? e) =>
+              e is AssertionError &&
+              e.message ==
+                  'Only one of uniformTypeIdentifiers or macUTIs can be non-null')));
```
Comment (ditman) [inline packages/file_selector/file_selector_platform_interface/test/x_type_group_test.dart:125]: I'm not a fan of asserting the text of the assertion error, I'd just do a `expect(closure, throwsAssertionError)` (or `throwsA(isA<AssertionError>())`)

**Review by ditman — APPROVED** [review]
Thanks for the change, let's land this! My comment is totally optional, I normally don't asssert error messages; admittedly this implementation is more strict :)
